// jsdoctypeparser/lib/TypeDictionary.js
'use strict';

/**
 * Namespace for type dictionaries.
 * @type {Object<string, boolean>}
 */
var TypeDictionary = {
  Array: true,
  Boolean: true,
  Date: true,
  Document: true,
  Element: true,
  Error: true,
  Function: true,
  Image: true,
  Infinity: true,
  Iterator: true,
  NaN: true,
  Null: true,
  Number: true,
  Object: true,
  Option: true,
  RegExp: true,
  String: true,
  Undefined: true,
  Window: true,

  EvalError: true,
  InternalError: true,
  RangeError: true,
  ReferenceError: true,
  StopIteration: true,
  SyntaxError: true,
  TypeError: true,
  UriError: true,

  Attr: true,
  CDATASection: true,
  CharacterData: true,
  Comment: true,
  console: true,
  DocumentFragment: true,
  DocumentType: true,
  DomConfiguration: true,
  DOMError: true,
  DOMErrorHandler: true,
  DOMException: true,
  DOMImplementation: true,
  DOMImplementationList: true,
  DOMImplementationRegistry: true,
  DOMImplementationSource: true,
  DOMLocator: true,
  DOMObject: true,
  DOMRequest: true,
  DOMString: true,
  DOMStringList: true,
  DOMTimeStamp: true,
  DOMUserData: true,
  Entity: true,
  EntityReference: true,
  MediaQueryList: true,
  MediaQueryListListener: true,
  NameList: true,
  NamedNodeMap: true,
  Node: true,
  NodeFilter: true,
  NodeIterator: true,
  NodeList: true,
  Notation: true,
  Plugin: true,
  PluginArray: true,
  ProcessingInstruction: true,
  SharedWorker: true,
  Text: true,
  TimeRanges: true,
  Treewalker: true,
  TypeInfo: true,
  UserDataHandler: true,
  Worker: true,
  WorkerGlobalScope: true,

  HTMLDocument: true,
  HTMLElement: true,
  HTMLAnchorElement: true,
  HTMLAppletElement: true,
  HTMLAudioElement: true,
  HTMLAreaElement: true,
  HTMLBaseElement: true,
  HTMLBaseFontElement: true,
  HTMLBodyElement: true,
  HTMLBRElement: true,
  HTMLButtonElement: true,
  HTMLCanvasElement: true,
  HTMLDataListElement: true,
  HTMLDirectoryElement: true,
  HTMLDivElement: true,
  HTMLDListElement: true,
  HTMLEmbedElement: true,
  HTMLFieldSetElement: true,
  HTMLFontElement: true,
  HTMLFormElement: true,
  HTMLFrameElement: true,
  HTMLFrameSetElement: true,
  HTMLHeadElement: true,
  HTMLHeadingElement: true,
  HTMLHtmlElement: true,
  HTMLHRElement: true,
  HTMLIFrameElement: true,
  HTMLImageElement: true,
  HTMLInputElement: true,
  HTMLKeygenElement: true,
  HTMLLabelElement: true,
  HTMLLIElement: true,
  HTMLLinkElement: true,
  HTMLMapElement: true,
  HTMLMenuElement: true,
  HTMLMetaElement: true,
  HTMLMeterElement: true,
  HTMLModElement: true,
  HTMLObjectElement: true,
  HTMLOListElement: true,
  HTMLOptGroupElement: true,
  HTMLOptionElement: true,
  HTMLOutputElement: true,
  HTMLParagraphElement: true,
  HTMLParamElement: true,
  HTMLPreElement: true,
  HTMLProgressElement: true,
  HTMLQuoteElement: true,
  HTMLScriptElement: true,
  HTMLSelectElement: true,
  HTMLSourceElement: true,
  HTMLSpanElement: true,
  HTMLStyleElement: true,
  HTMLTableElement: true,
  HTMLTableCaptionElement: true,
  HTMLTableCellElement: true,
  HTMLTableDataCellElement: true,
  HTMLTableHeaderCellElement: true,
  HTMLTableColElement: true,
  HTMLTableRowElement: true,
  HTMLTableSectionElement: true,
  HTMLTextAreaElement: true,
  HTMLTimeElement: true,
  HTMLTitleElement: true,
  HTMLTrackElement: true,
  HTMLUListElement: true,
  HTMLUnknownElement: true,
  HTMLVideoElement: true,
  CanvasRenderingContext2D: true,
  CanvasGradient: true,
  CanvasPattern: true,
  TextMetrics: true,
  ImageData: true,
  CanvasPixelArray: true,

  SVGElement: true,
  SVGAElement: true,
  SVGAltGlyphElement: true,
  SVGAltGlyphDefElement: true,
  SVGAltGlyphItemElement: true,
  SVGAnimationElement: true,
  SVGAnimateElement: true,
  SVGAnimateColorElement: true,
  SVGAnimateMotionElement: true,
  SVGAnimateTransformElement: true,
  SVGSetElement: true,
  SVGCircleElement: true,
  SVGClipPathElement: true,
  SVGColorProfileElement: true,
  SVGCursorElement: true,
  SVGDefsElement: true,
  SVGDescElement: true,
  SVGEllipseElement: true,
  SVGFilterElement: true,
  SVGFilterPrimitiveStandardAttributes: true,
  SVGFEBlendElement: true,
  SVGFEColorMatrixElement: true,
  SVGFEComponentTransferElement: true,
  SVGFECompositeElement: true,
  SVGFEConvolveMatrixElement: true,
  SVGFEDiffuseLightingElement: true,
  SVGFEDisplacementMapElement: true,
  SVGFEDistantLightElement: true,
  SVGFEFloodElement: true,
  SVGFEGaussianBlurElement: true,
  SVGFEImageElement: true,
  SVGFEMergeElement: true,
  SVGFEMergeNodeElement: true,
  SVGFEMorphologyElement: true,
  SVGFEOffsetElement: true,
  SVGFEPointLightElement: true,
  SVGFESpecularLightingElement: true,
  SVGFESpotLightElement: true,
  SVGFETileElement: true,
  SVGFETurbulenceElement: true,
  SVGComponentTransferFunctionElement: true,
  SVGFEFuncRElement: true,
  SVGFEFuncGElement: true,
  SVGFEFuncBElement: true,
  SVGFEFuncAElement: true,
  SVGFontElement: true,
  SVGFontFaceElement: true,
  SVGFontFaceFormatElement: true,
  SVGFontFaceNameElement: true,
  SVGFontFaceSrcElement: true,
  SVGFontFaceUriElement: true,
  SVGForeignObjectElement: true,
  SVGGElement: true,
  SVGGlyphElement: true,
  SVGGlyphRefElement: true,
  SVGGradientElement: true,
  SVGLinearGradientElement: true,
  SVGRadialGradientElement: true,
  SVGHKernElement: true,
  SVGImageElement: true,
  SVGLineElement: true,
  SVGMarkerElement: true,
  SVGMaskElement: true,
  SVGMetadataElement: true,
  SVGMissingGlyphElement: true,
  SVGMPathElement: true,
  SVGPathElement: true,
  SVGPatternElement: true,
  SVGPolylineElement: true,
  SVGPolygonElement: true,
  SVGRectElement: true,
  SVGScriptElement: true,
  SVGStopElement: true,
  SVGStyleElement: true,
  SVGSVGElement: true,
  SVGSwitchElement: true,
  SVGSymbolElement: true,
  SVGTextElement: true,
  SVGTextPathElement: true,
  SVGTitleElement: true,
  SVGTRefElement: true,
  SVGTSpanElement: true,
  SVGUseElement: true,
  SVGViewElement: true,
  SVGVKernElement: true,
};


// Exports the constructor.
module.exports = TypeDictionary;
